tests/msc3030_test.go

```diff
@@ -0,0 +1,147 @@
+// +build msc3030
+
+// This file contains tests for a jump to date API endpoint,
+// currently experimental feature defined by MSC3030, which you can read here:
+// https://github.com/matrix-org/matrix-doc/pull/3030
+
+package tests
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/tidwall/gjson"
+)
+
+func TestJumpToDateEndpoint(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintFederationTwoLocalOneRemote)
+	defer deployment.Destroy(t)
+
+	// Create the normal user which will send messages in the room
+	userID := "@alice:hs1"
+	alice := deployment.Client(t, "hs1", userID)
+
+	roomID := alice.CreateRoom(t, map[string]interface{}{})
+	alice.JoinRoom(t, roomID, nil)
+
+	timeBeforeEventA := time.Now()
+	eventAID := alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Message A",
+		},
+	})
+	eventBID := alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Message A",
+		},
+	})
+	timeAfterEventB := time.Now()
+
+	t.Run("parallel", func(t *testing.T) {
+		t.Run("should find event after given timestmap", func(t *testing.T) {
+			checkEventisReturnedForTime(t, alice, roomID, timeBeforeEventA, eventAID)
+		})
+
+		t.Run("should find event before given timestmap", func(t *testing.T) {
+			checkEventisReturnedForTime(t, alice, roomID, timeAfterEventB, eventBID)
+		})
+	})
+}
+
+func checkEventisReturnedForTime(t *testing.T, c *client.CSAPI, roomID string, givenTime time.Time, expectedEventId string) {
+	t.Helper()
+
+	givenTimestamp := makeTimestampFromTime(givenTime)
+	timestampString := strconv.FormatInt(givenTimestamp, 10)
+	timestampToEventRes := c.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "timestamp_to_event"}, client.WithContentType("application/json"), client.WithQueries(url.Values{
+		"ts": []string{timestampString},
+	}))
+	timestampToEventResBody := client.ParseJSON(t, timestampToEventRes)
+
+	actualEventIdRes := gjson.GetBytes(timestampToEventResBody, "event_id")
+	actualEventId := actualEventIdRes.String()
+
+	if actualEventId != expectedEventId {
+		debugMessageList := getDebugMessageListFromMessagesResponse(t, c, roomID, expectedEventId, actualEventId, givenTimestamp)
+		t.Fatalf(
+			"Expected to see %s given %s but received %s\n%s",
+			decorateStringWithAnsiColor(expectedEventId, AnsiColorGreen),
+			decorateStringWithAnsiColor(timestampString, AnsiColorYellow),
+			decorateStringWithAnsiColor(actualEventId, AnsiColorRed),
+			debugMessageList,
+		)
```

Comment:
This will give us a nice debug statement when the test fails. It gives us the full history of messages in the room with color coding to show the expected and actual event alongside where the given timestamp fits into the list of events.

<img width="1299" alt="Screen Shot 2021-07-28 at 2 17 39 AM" src="https://user-images.githubusercontent.com/558581/127280520-b4444684-91bd-4faa-9b87-d299408b26cf.png">


```
    msc3030_test.go:55: Expected to see $1FLuxa0W9wsuZADc50Ze4p2lBkBVTdvo2_dNkTxZeEw given 1627456572273 but received $Ip9mMxOYoxbF6QGRIx9ikceYyoCcDE1omR9dGDlxiMU
        -- givenTimestamp=1627456572273 --
        $1FLuxa0W9wsuZADc50Ze4p2lBkBVTdvo2_dNkTxZeEw (1627456572237) - m.room.message
        $XcV2Iv6-g_NkbHUQGeK0MT_nt6PFr0P--XD3qqjfzC4 (1627456572168) - m.room.message
        $SSMS9huj-jCQbWra-fk_u7VgYAuBZBcrU5btXF31eMo (1627456572116) - m.room.history_visibility
        $kI6FjoqF3E2YhZKO9w2ulqJNooeHF6iYdZ17F85sXRA (1627456572093) - m.room.join_rules
        $kCzsgUT298MOvg1LzyzFBSKIVu-_GRmmcGvPah-0pxU (1627456572064) - m.room.power_levels
        $6vmslxv6rY9rehCQ8SBCsUbBYB0r_LjOpC7KHrOimt0 (1627456572030) - m.room.member
        $Ip9mMxOYoxbF6QGRIx9ikceYyoCcDE1omR9dGDlxiMU (1627456571972) - m.room.create
```

Reply:
Updated output:

<img width="1277" alt="" src="https://user-images.githubusercontent.com/558581/142157993-bcbea9c4-4b42-47e5-a86a-64939aded535.png">

```
    msc3030_test.go:71: Expected to see $eUp_uwB1OOKpf8PS2gjKeBb8idZJdrxjZJgJOABj6gE given 1637133372333 but received $cj3BEfTs6V9q9xMYmUqZtBUJmiGU6jKx3kGTDLZI9e4
        (oldest)
        $_kqjKsk3UnzD8Iz-OL8foxXujrMN3DVTuwtkwSq4Egg (1637133372086) - m.room.create
        $SqcsUcy0X1lPpZXiPAdD3kXaVyp69YsE36dZEnOSwyo (1637133372140) - m.room.member
        $l1wm_7MxNs8XxvR3l8HRWqNFjuDf_-lk4HDurQycQC4 (1637133372170) - m.room.power_levels
        $RKl-ad2a0akyd-TZqmQDhdW50wIBny6mEDZPt3JL13Q (1637133372209) - m.room.join_rules
        $cj3BEfTs6V9q9xMYmUqZtBUJmiGU6jKx3kGTDLZI9e4 (1637133372234) - m.room.history_visibility
        $CwJhk21z8KDgsjIpzvXKTizQzRDIUd05eXAwGNvAruc (1637133372269) - m.room.message
        $eUp_uwB1OOKpf8PS2gjKeBb8idZJdrxjZJgJOABj6gE (1637133372324) - m.room.message
        -- givenTimestamp=1637133372333 --
        $-MtFS4tnzFYdrZx3eRVxp7iUyw14e08Y11xlZIQIS8E (1637133373030) - m.room.member
        (newest)
```

Reply:
I would recommend adding `(expected)` and `(received)` bits after the events in the timeline, it's not as easily visible when without colours.

Reply:
How does this look? 🙂

<img width="1281" alt="" src="https://user-images.githubusercontent.com/558581/144550094-74580d6a-e879-42ea-a6f2-fd6b169ae426.png">


```
    msc3030_test.go:37: Expected to see $XQvueEX5r9wLDhH15tXm_weWOQ-JlTU4o2JWkkwnYC0 given 1638506547868 but received $XQvueEX5r9wLDhH15tXm_weWOQ-JlTU4o2JWkkwnYC0
                   -- oldest events --
          (actual) $lapGquM-EboC-7FTGMYKw8LheBYqjobb6j0M2AusS3k (1638506547354) - m.room.create
                   $FkBUU9NGgvYJwQ_ofXsNm-omUTO6mkUnGBFj2kOiJJ8 (1638506547442) - m.room.member
                   $faI0p28WHp-8fnUskgDrngBpXVkxWgzTFOEuJ-a0JHk (1638506547536) - m.room.power_levels
                   $3AYYdSHWG-NvWoIoDmtEDXl073yGFpz9aJn7ftSwJlg (1638506547695) - m.room.join_rules
                   $_hHlabu_rAycc78meQX8Y-M-IBv_PQth0X5Ybt41Jp4 (1638506547792) - m.room.history_visibility
                   -- givenTimestamp=1638506547868 --
        (expected) $XQvueEX5r9wLDhH15tXm_weWOQ-JlTU4o2JWkkwnYC0 (1638506547919) - m.room.message
                   $sU76QDNGmARqRCgGMoZ6vwtq9JvccO7nWyLD-qNp3hs (1638506548108) - m.room.message
                   -- newest events --
```

Reply:
I don't think "actual" reflects it as well as "expected", it's fine enough as-is, but the wording might trip up.

Reply:
It's both `actual`(what we received) and `expected`(what we should see) which matches a lot of testing assertion functions,

 - https://pkg.go.dev/github.com/stretchr/testify/assert#Equal
 - https://nodejs.org/api/assert.html#assertstrictequalactual-expected-message

Reply:
Ah, then that makes more sense, then this is pretty much resolved, thanks!

Reply:
If we're being pedantic with naming, it should be "got" and "want" as per our Go overlords. The stdlib uses that terminology basically everywhere and it's a recommendation on most articles about testing in Go.

Reply:
How's this?

<img width="1150" alt="" src="https://user-images.githubusercontent.com/558581/146620870-e9d4f810-41c5-4d96-b5a1-81ed06dad825.png">